// @zup-it/nimbus-backend-core/jsx/structure/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Switch = exports.nimbusFragmentFactory = exports.ForEach = exports.Else = exports.Then = exports.If = void 0;
var condition_1 = require("./condition");
Object.defineProperty(exports, "If", { enumerable: true, get: function () { return condition_1.If; } });
Object.defineProperty(exports, "Then", { enumerable: true, get: function () { return condition_1.Then; } });
Object.defineProperty(exports, "Else", { enumerable: true, get: function () { return condition_1.Else; } });
var foreach_1 = require("./foreach");
Object.defineProperty(exports, "ForEach", { enumerable: true, get: function () { return foreach_1.ForEach; } });
var fragment_1 = require("./fragment");
Object.defineProperty(exports, "nimbusFragmentFactory", { enumerable: true, get: function () { return fragment_1.nimbusFragmentFactory; } });
var switch_1 = require("./switch");
Object.defineProperty(exports, "Switch", { enumerable: true, get: function () { return switch_1.Switch; } });
