// @zup-it/nimbus-backend-core/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getGlobalState = exports.componentValidation = exports.Switch = exports.ForEach = exports.Else = exports.Then = exports.If = exports.setFragmentFactory = exports.NimbusJSX = exports.Operation = exports.setupHotReloading = exports.isDynamicExpression = exports.serialize = exports.createState = exports.createStateNode = exports.createAction = exports.RootState = exports.Component = void 0;
var component_1 = require("./model/component");
Object.defineProperty(exports, "Component", { enumerable: true, get: function () { return component_1.Component; } });
var root_state_1 = require("./model/state/root-state");
Object.defineProperty(exports, "RootState", { enumerable: true, get: function () { return root_state_1.RootState; } });
var action_1 = require("./model/action");
Object.defineProperty(exports, "createAction", { enumerable: true, get: function () { return action_1.createAction; } });
var state_node_1 = require("./model/state/state-node");
Object.defineProperty(exports, "createStateNode", { enumerable: true, get: function () { return state_node_1.createStateNode; } });
var root_state_2 = require("./model/state/root-state");
Object.defineProperty(exports, "createState", { enumerable: true, get: function () { return root_state_2.createState; } });
var serializer_1 = require("./serializer");
Object.defineProperty(exports, "serialize", { enumerable: true, get: function () { return serializer_1.serialize; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "isDynamicExpression", { enumerable: true, get: function () { return utils_1.isDynamicExpression; } });
Object.defineProperty(exports, "setupHotReloading", { enumerable: true, get: function () { return utils_1.setupHotReloading; } });
var operation_1 = require("./model/operation");
Object.defineProperty(exports, "Operation", { enumerable: true, get: function () { return operation_1.Operation; } });
var jsx_1 = require("./jsx");
Object.defineProperty(exports, "NimbusJSX", { enumerable: true, get: function () { return jsx_1.NimbusJSX; } });
Object.defineProperty(exports, "setFragmentFactory", { enumerable: true, get: function () { return jsx_1.setFragmentFactory; } });
var structure_1 = require("./jsx/structure");
Object.defineProperty(exports, "If", { enumerable: true, get: function () { return structure_1.If; } });
Object.defineProperty(exports, "Then", { enumerable: true, get: function () { return structure_1.Then; } });
Object.defineProperty(exports, "Else", { enumerable: true, get: function () { return structure_1.Else; } });
Object.defineProperty(exports, "ForEach", { enumerable: true, get: function () { return structure_1.ForEach; } });
Object.defineProperty(exports, "Switch", { enumerable: true, get: function () { return structure_1.Switch; } });
__exportStar(require("./operations"), exports);
var validation_1 = require("./validation");
Object.defineProperty(exports, "componentValidation", { enumerable: true, get: function () { return validation_1.componentValidation; } });
var global_state_1 = require("./global-state");
Object.defineProperty(exports, "getGlobalState", { enumerable: true, get: function () { return global_state_1.getGlobalState; } });
